tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -4215,15 +4306,15 @@ private[optimizer] object OptimizerCore {
       RefinedType(base, isExact, isNullable, AllocationSite.Anonymous)
 
     def apply(tpe: Type): RefinedType = tpe match {
-      case IntType | FloatType | DoubleType =>
-        RefinedType(tpe, isExact = false, isNullable = false)
-      case BooleanType | StringType | UndefType | NothingType |
-          _:RecordType | NoType =>
-        RefinedType(tpe, isExact = true, isNullable = false)
+      case AnyType | ClassType(_) | ArrayType(_) =>
+        RefinedType(tpe, isExact = false, isNullable = true)
       case NullType =>
         RefinedType(tpe, isExact = true, isNullable = true)
-      case _ =>
-        RefinedType(tpe, isExact = false, isNullable = true)
+      case NothingType | UndefType | BooleanType | CharType | ByteType |
+          LongType | StringType | NoType =>
+        RefinedType(tpe, isExact = true, isNullable = false)
+      case ShortType | IntType | FloatType | DoubleType | RecordType(_) =>
+        RefinedType(tpe, isExact = false, isNullable = false)
```

Comment:
I also thought so, and the CI saved me from my mistake. In the *type system* they're not subtypes, but *at run-time* there's another story: a small `int` will still answer `true` to `x.isInstanceOf[Byte]`. Since `isExact` is used to optimize away `isInstanceOf`s that answer `false`, we need those to be considered non-exact.

In fact, I think now even `ByteType` needs to be non-exact :-s Right now I think I broke `(1.toByte: Any).isInstanceOf[Int]`. It would answer `true` at run-time but the optimizer will replace it with `false`.